File: ocaml/plugins/allplugins.ml
(* THE FULL SET - every plugin this library ships, in one module.

   It exists for the callers that genuinely want all ten kinds: the CLI,
   the conformance suite, an app whose chain is decided at run time.

     let secrets = Sekreto.sekreto ~plugins:(Allplugins.all ()) chain

   IT IS ALSO THE THING TO AVOID IF YOU CARE ABOUT SIZE. Naming this module
   links every plugin - AWS request signing, seven HTTP vault clients, the
   TLS binding and the OpenSSL beneath it - which is the cost the
   core/plugin split exists to remove. A lean consumer names the kinds it
   actually configures, each from its own module:

     let secrets = Sekreto.sekreto ~plugins:[ Hashicorp.plugin () ] chain

   ocamlopt links a compilation unit only when something references it, so
   the difference is in the binary and not merely in the source. *)

(* Built, not held: every entry is a fresh definition, so two chains never
   share one and nothing is built at load time. *)
let all () : Defs.definition list =
  [
    Hashicorp.plugin ();
    Boru.plugin ();
    Aws.secretsplugin ();
    Aws.paramsplugin ();
    Gcpsecrets.plugin ();
    Azuresecrets.plugin ();
    Onepassword.plugin ();
    Doppler.plugin ();
    Infisical.plugin ();
    Secretspec.plugin ();
  ]

File: haskell/plugins/AllPlugins.hs
-- | THE FULL SET - every plugin this port ships, in one import.
--
-- It exists for the callers that genuinely want all ten kinds: the CLI,
-- the conformance suite, an app whose chain is decided at run time.
--
-- > secrets <- sekreto emptyoptions {optplugins = allplugins, optproviders = chain}
--
-- IT IS ALSO THE THING TO AVOID IF YOU CARE ABOUT SIZE. Importing this
-- module compiles every plugin into the program - AWS request signing,
-- seven HTTP vault clients and the OpenSSL binding under them - which is
-- the cost the core/plugin split exists to remove. A lean consumer
-- imports the kinds it actually configures, each from its own module:
--
-- > import Hashicorp (hashicorp)
-- > secrets <- sekreto emptyoptions {optplugins = [hashicorp], optproviders = chain}

module AllPlugins
  ( allplugins,
    awsparams,
    awssecrets,
    azuresecrets,
    boru,
    doppler,
    gcpsecrets,
    hashicorp,
    infisical,
    onepassword,
    secretspec,
  )
where

import Aws (awsparams, awssecrets)
import Azuresecrets (azuresecrets)
import Boru (boru)
import Defs (Definition)
import Doppler (doppler)
import Gcpsecrets (gcpsecrets)
import Hashicorp (hashicorp)
import Infisical (infisical)
import Onepassword (onepassword)
import Secretspec (secretspec)

-- | Every plugin definition this port ships, in a fresh list and in the
-- order the canonical lists them.
allplugins :: [Definition]
allplugins =
  [ hashicorp,
    boru,
    awssecrets,
    awsparams,
    gcpsecrets,
    azuresecrets,
    onepassword,
    doppler,
    infisical,
    secretspec
  ]
